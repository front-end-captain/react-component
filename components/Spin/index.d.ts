import * as React from "react";

interface SpinProps {
  prefixCls?: string;
  className?: string;
  style?: React.CSSProperties;
  spinning?: boolean;
  size?: "small" | "default" | "large";
  tip?: React.ReactNode;
  indicator?: React.ReactElement<any>;

  // TODO: 延迟显示加载效果的时间
  // delay?: number;
}

declare const Spin: (props: SpinProps) => React.ReactNode;

export default Spin;
