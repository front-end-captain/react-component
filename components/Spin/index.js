import Spin from "./spin.jsx";

export default Spin;
