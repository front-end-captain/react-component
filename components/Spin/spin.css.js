import styled from "styled-components";

import { primaryColor } from "./../_style/vars.js";
import { skySpinRotate } from "./../_style/animate.js";

const Wrapper = styled.div`
  position: relative;

  .${(props) => props.prefixCls}-wrapper {
    position: absolute;
    display: flex;
    align-items: center;
    justify-content: center;
    height: 100%;
    width: 100%;
    z-index: 4;
  }

  .${(props) => props.prefixCls}-child-container {
    position: relative;
    transition: opacity 0.3s;
  }

  .${(props) => props.prefixCls}-child-blur {
    pointer-events: none;
    user-select: none;
    overflow: hidden;
    opacity: 0.5;
  }

  .${(props) => props.prefixCls}-child-blur:after {
    display: block;
    opacity: 0.4;
    pointer-events: auto;
  }

  .${(props) => props.prefixCls}-child-container:after {
    display: none;
    content: "";
    position: absolute;
    left: 0;
    right: 0;
    top: 0;
    bottom: 0;
    background: #fff;
    pointer-events: none;
    transition: all 0.3s;
    height: 100%;
    width: 100%;
    z-index: 10;
    filter: blur(1px);
  }
`;

const SpinWrapper = styled.div`
  display: inline-block;
  text-align: center;

  .${(props) => props.prefixCls}-tip {
    display: block;
    color: ${primaryColor};
    font-size: 14px;
    margin-top: 4px;
  }

  &.${(props) => props.prefixCls}-spinning {
    .${(props) => props.prefixCls}-spinning-icon {
      display: inline-block;
      vertical-align: middle;
      color: ${primaryColor};
      font-size: 20px;
      width: 20px;
      height: 20px;
      animation: ${skySpinRotate} 1.2s linear infinite;
    }
  }

  &.${(props) => props.prefixCls}-small {
    .${(props) => props.prefixCls}-spinning-icon {
      font-size: 14px;
      width: 14px;
      height: 14px;
    }
  }

  &.${(props) => props.prefixCls}-large {
    .${(props) => props.prefixCls}-spinning-icon {
      font-size: 32px;
      width: 32px;
      height: 32px;
    }
  }
`;
export { SpinWrapper, Wrapper };
