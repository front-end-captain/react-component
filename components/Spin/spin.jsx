import React, { isValidElement, cloneElement } from "react";
import PropTypes from "prop-types";
import classNames from "classnames";

import { PointLoadingIcon } from "./../Icon/index.jsx";

import { SpinWrapper, Wrapper } from "./spin.css.js";

const getIndicator = (element, iconClassName) => {
  if (isValidElement(element)) {
    return cloneElement(element, { className: iconClassName });
  }

  return <PointLoadingIcon className={iconClassName} />;
};

const Spin = (props) => {
  const { prefixCls, className, style, spinning, tip, indicator, size, children } = props;

  const iconClassName = `${prefixCls}-spinning-icon`;
  const wrapClassName = classNames(`${prefixCls}-wrapper`, className, `${prefixCls}-${size}`, {
    [`${prefixCls}-spinning`]: spinning,
  });
  const childContainerClassName = classNames(`${prefixCls}-child-container`, {
    [`${prefixCls}-child-blur`]: spinning,
  });

  const spinContainer = (
    <SpinWrapper className={wrapClassName} style={style} prefixCls={prefixCls}>
      {getIndicator(indicator, iconClassName)}
      {tip && <span className={`${prefixCls}-tip`}>{tip}</span>}
    </SpinWrapper>
  );

  if (children !== undefined) {
    return (
      <Wrapper className={classNames(`${prefixCls}-nested-wrapper`)} prefixCls={prefixCls}>
        {spinning && spinContainer}
        <div className={childContainerClassName}>{children}</div>
      </Wrapper>
    );
  }

  return spinContainer;
};

Spin.defaultProps = {
  prefixCls: "sky-spin",
  size: "default",
  spinning: true,
};

Spin.propTypes = {
  prefixCls: PropTypes.string,
  className: PropTypes.string,
  style: PropTypes.objectOf(PropTypes.string),
  tip: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
  indicator: PropTypes.element,
  size: PropTypes.oneOf(["small", "default", "large"]),
  spinning: PropTypes.bool,
};

export default Spin;
