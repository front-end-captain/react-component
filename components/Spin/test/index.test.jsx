import React from "react";
import { render, shallow } from "enzyme";

import Spin from "./../index.js";

describe("<Spin />", () => {
  it("should only affect the Spin element when set style to a nested <Spin>xxx</Spin>", () => {
    const wrapper = shallow(
      <Spin style={{ background: "red" }}>
        <div>content</div>
      </Spin>,
    );

    expect(
      wrapper
        .find(".sky-spin-nested-wrapper")
        .at(0)
        .prop("style"),
    ).toBeFalsy();

    expect(
      wrapper
        .find(".sky-spin-wrapper")
        .at(0)
        .prop("style").background,
    ).toBe("red");
  });

  it("should render custom indicator when it's set indicator", () => {
    const customIndicator = <div className="custom-indicator" />;
    const wrapper = render(<Spin indicator={customIndicator} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should be controlled by spinning", () => {
    const wrapper = shallow(<Spin spinning={false} />);
    expect(wrapper.find(".sky-spin-spinning").length === 1);

    wrapper.setProps({ spinning: true });
    expect(wrapper.find(".sky-spin-spinning").length === 0);
  });
});
