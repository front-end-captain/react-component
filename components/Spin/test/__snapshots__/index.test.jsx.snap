// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Spin /> should render custom indicator when it's set indicator 1`] = `
<div
  class="sky-spin-wrapper sky-spin-default sky-spin-spinning spincss__SpinWrapper-sc-194o6xu-1 ekEaJL"
>
  <div
    class="sky-spin-spinning-icon"
  />
</div>
`;
